README.md
# Uptime Kuma

<a target="_blank" href="https://github.com/louislam/uptime-kuma"><img src="https://img.shields.io/github/stars/louislam/uptime-kuma" /></a> <a target="_blank" href="https://hub.docker.com/r/louislam/uptime-kuma"><img src="https://img.shields.io/docker/pulls/louislam/uptime-kuma" /></a> <a target="_blank" href="https://hub.docker.com/r/louislam/uptime-kuma"><img src="https://img.shields.io/docker/v/louislam/uptime-kuma/latest?label=docker%20image%20ver." /></a> <a target="_blank" href="https://github.com/louislam/uptime-kuma"><img src="https://img.shields.io/github/last-commit/louislam/uptime-kuma" /></a>  <a target="_blank" href="https://opencollective.com/uptime-kuma"><img src="https://opencollective.com/uptime-kuma/total/badge.svg?label=Open%20Collective%20Backers&color=brightgreen" /></a>
[![GitHub Sponsors](https://img.shields.io/github/sponsors/louislam?label=GitHub%20Sponsors)](https://github.com/sponsors/louislam) <a href="https://weblate.kuma.pet/projects/uptime-kuma/uptime-kuma/">
<img src="https://weblate.kuma.pet/widgets/uptime-kuma/-/svg-badge.svg" alt="Translation status" />
</a>

<div align="center" width="100%">
    <img src="./public/icon.svg" width="128" alt="" />
</div>

Uptime Kuma is an easy-to-use self-hosted monitoring tool.

<img src="https://user-images.githubusercontent.com/1336778/212262296-e6205815-ad62-488c-83ec-a5b0d0689f7c.jpg" width="700" alt="" />

## 🥔 Live Demo

Try it!

- Tokyo Demo Server: https://demo.uptime.kuma.pet (Sponsored by [Uptime Kuma Sponsors](https://github.com/louislam/uptime-kuma#%EF%B8%8F-sponsors))

It is a temporary live demo, all data will be deleted after 10 minutes. Use the one that is closer to you, but I suggest that you should install and try it out for the best demo experience.

## ⭐ Features

* Monitoring uptime for HTTP(s) / TCP / HTTP(s) Keyword / Ping / DNS Record / Push / Steam Game Server / Docker Containers
* Fancy, Reactive, Fast UI/UX
* Notifications via Telegram, Discord, Gotify, Slack, Pushover, Email (SMTP), and [90+ notification services, click here for the full list](https://github.com/louislam/uptime-kuma/tree/master/src/components/notifications)
* 20 second intervals
* [Multi Languages](https://github.com/louislam/uptime-kuma/tree/master/src/lang)
* Multiple status pages
* Map status pages to specific domains
* Ping chart
* Certificate info
* Proxy support
* 2FA support

## 🔧 How to Install

### 🐳 Docker

```bash
docker run -d --restart=always -p 3001:3001 -v uptime-kuma:/app/data --name uptime-kuma louislam/uptime-kuma:1
```

⚠️ Please use a **local volume** only. Other types such as NFS are not supported.

Uptime Kuma is now running on http://localhost:3001

### 💪🏻 Non-Docker

Required Tools: 
- [Node.js](https://nodejs.org/en/download/) >= 14
- [Git](https://git-scm.com/downloads) 
- [pm2](https://pm2.keymetrics.io/) - For running Uptime Kuma in the background

```bash
# Update your npm to the latest version
npm install npm -g

git clone https://github.com/louislam/uptime-kuma.git
cd uptime-kuma
npm run setup

# Option 1. Try it
node server/server.js

# (Recommended) Option 2. Run in background using PM2
# Install PM2 if you don't have it: 
npm install pm2 -g && pm2 install pm2-logrotate

# Start Server
pm2 start server/server.js --name uptime-kuma


```
Uptime Kuma is now running on http://localhost:3001

More useful PM2 Commands

```bash
# If you want to see the current console output
pm2 monit

# If you want to add it to startup
pm2 save && pm2 startup
```

### Advanced Installation

If you need more options or need to browse via a reverse proxy, please read:

https://github.com/louislam/uptime-kuma/wiki/%F0%9F%94%A7-How-to-Install

## 🆙 How to Update

Please read:

https://github.com/louislam/uptime-kuma/wiki/%F0%9F%86%99-How-to-Update

## 🆕 What's Next?

I will mark requests/issues to the next milestone.

https://github.com/louislam/uptime-kuma/milestones

Project Plan:

https://github.com/users/louislam/projects/4/views/1

## ❤️ Sponsors

Thank you so much! (GitHub Sponsors will be updated manually. OpenCollective sponsors will be updated automatically, the list will be cached by GitHub though. It may need some time to be updated)

<img src="https://uptime.kuma.pet/sponsors?v=6" alt />

## 🖼 More Screenshots

Light Mode:

<img src="https://uptime.kuma.pet/img/light.jpg" width="512" alt="" />

Status Page:

<img src="https://user-images.githubusercontent.com/1336778/134628766-a3fe0981-0926-4285-ab46-891a21c3e4cb.png" width="512" alt="" />

Settings Page:

<img src="https://louislam.net/uptimekuma/2.jpg" width="400" alt="" />

Telegram Notification Sample:

<img src="https://louislam.net/uptimekuma/3.jpg" width="400" alt="" />

## Motivation

* I was looking for a self-hosted monitoring tool like "Uptime Robot", but it is hard to find a suitable one. One of the close ones is statping. Unfortunately, it is not stable and no longer maintained.
* Want to build a fancy UI.
* Learn Vue 3 and vite.js.
* Show the power of Bootstrap 5.
* Try to use WebSocket with SPA instead of REST API.
* Deploy my first Docker image to Docker Hub.

If you love this project, please consider giving me a ⭐.

## 🗣️ Discussion

### Issues Page

You can discuss or ask for help in [issues](https://github.com/louislam/uptime-kuma/issues).

### Subreddit

My Reddit account: [u/louislamlam](https://reddit.com/u/louislamlam).  
You can mention me if you ask a question on Reddit.
[r/Uptime kuma](https://www.reddit.com/r/UptimeKuma/)

## Contribute

### Test Pull Requests

There are a lot of pull requests right now, but I don't have time to test them all.

If you want to help, you can check this:
https://github.com/louislam/uptime-kuma/wiki/Test-Pull-Requests

### Test Beta Version

Check out the latest beta release here: https://github.com/louislam/uptime-kuma/releases

### Bug Reports / Feature Requests
If you want to report a bug or request a new feature, feel free to open a [new issue](https://github.com/louislam/uptime-kuma/issues).

### Translations
If you want to translate Uptime Kuma into your language, please visit [Weblate Readme](https://github.com/louislam/uptime-kuma/blob/master/src/lang/README.md).

Feel free to correct my grammar in this README, source code, or wiki, as my mother language is not English and my grammar is not that great.

### Create Pull Requests
If you want to modify Uptime Kuma, please read this guide and follow the rules here: https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md


CHANGELOG.md
## 1.19.7 (2023-1-29)

- chore: wrap as blocklet


install.sh
# install.sh is generated by ./extra/install.batsh, do not modify it directly.
# "npm run compile-install-script" to compile install.sh
# The command is working on Windows PowerShell and Docker for Windows only.
# curl -o kuma_install.sh https://raw.githubusercontent.com/louislam/uptime-kuma/master/install.sh && sudo bash kuma_install.sh
"echo" "-e" "====================="
"echo" "-e" "Uptime Kuma Installer"
"echo" "-e" "====================="
"echo" "-e" "Supported OS: CentOS 7/8, Ubuntu >= 16.04 and Debian"
"echo" "-e" "---------------------------------------"
"echo" "-e" "This script is designed for Linux and basic usage."
"echo" "-e" "For advanced usage, please go to https://github.com/louislam/uptime-kuma/wiki/Installation"
"echo" "-e" "---------------------------------------"
"echo" "-e" ""
"echo" "-e" "Local - Install Uptime Kuma in your current machine with git, Node.js 14 and pm2"
"echo" "-e" "Docker - Install Uptime Kuma Docker container"
"echo" "-e" ""
if [ "$1" != "" ]; then
  type="$1"
else
  "read" "-p" "Which installation method do you prefer? [DOCKER/local]: " "type"
fi
defaultPort="3001"
function checkNode {
  local _0
  nodeVersion=$(node -e 'console.log(process.versions.node.split(`.`)[0])')
  "echo" "-e" "Node Version: ""$nodeVersion"
  _0="12"
  if [ $(($nodeVersion < $_0)) == 1 ]; then
    "echo" "-e" "Error: Required Node.js 14"
    "exit" "1"  
fi
  if [ "$nodeVersion" == "12" ]; then
    "echo" "-e" "Warning: NodeJS ""$nodeVersion"" is not tested."  
fi
}
function deb {
  nodeCheck=$(node -v)
  apt --yes update
  if [ "$nodeCheck" != "" ]; then
    "checkNode" 
  else
    # Old nodejs binary name is "nodejs"
    check=$(nodejs --version)
    if [ "$check" != "" ]; then
      "echo" "-e" "Error: 'node' command is not found, but 'nodejs' command is found. Your NodeJS should be too old."
      exit 1    
fi
    curlCheck=$(curl --version)
    if [ "$curlCheck" == "" ]; then
      "echo" "-e" "Installing Curl"
      apt --yes install curl    
fi
    "echo" "-e" "Installing Node.js 14"
    curl -sL https://deb.nodesource.com/setup_14.x | bash - > log.txt
    apt --yes install nodejs
    node -v
    nodeCheckAgain=$(node -v)
    if [ "$nodeCheckAgain" == "" ]; then
      "echo" "-e" "Error during Node.js installation"
      exit 1    
fi
  fi
  check=$(git --version)
  if [ "$check" == "" ]; then
    "echo" "-e" "Installing Git"
    apt --yes install git  
fi
}
if [ "$type" == "local" ]; then
  defaultInstallPath="/opt/uptime-kuma"
  if [ -e "/etc/redhat-release" ]; then
    os=$("cat" "/etc/redhat-release")
    distribution="rhel"
  else
    if [ -e "/etc/issue" ]; then
      os=$(head -n1 /etc/issue | cut -f 1 -d ' ')
      if [ "$os" == "Ubuntu" ]; then
        distribution="ubuntu"      
fi
      if [ "$os" == "Debian" ]; then
        distribution="debian"      
fi    
fi
  fi
  arch=$(uname -i)
  "echo" "-e" "Your OS: ""$os"
  "echo" "-e" "Distribution: ""$distribution"
  "echo" "-e" "Arch: ""$arch"
  if [ "$3" != "" ]; then
    port="$3"
  else
    "read" "-p" "Listening Port [$defaultPort]: " "port"
    if [ "$port" == "" ]; then
      port="$defaultPort"    
fi
  fi
  if [ "$2" != "" ]; then
    installPath="$2"
  else
    "read" "-p" "Installation Path [$defaultInstallPath]: " "installPath"
    if [ "$installPath" == "" ]; then
      installPath="$defaultInstallPath"    
fi
  fi
  # CentOS
  if [ "$distribution" == "rhel" ]; then
    nodeCheck=$(node -v)
    if [ "$nodeCheck" != "" ]; then
      "checkNode" 
    else
      curlCheck=$(curl --version)
      if [ "$curlCheck" == "" ]; then
        "echo" "-e" "Installing Curl"
        yum -y -q install curl      
fi
      "echo" "-e" "Installing Node.js 14"
      curl -sL https://rpm.nodesource.com/setup_14.x | bash - > log.txt
      yum install -y -q nodejs
      node -v
      nodeCheckAgain=$(node -v)
      if [ "$nodeCheckAgain" == "" ]; then
        "echo" "-e" "Error during Node.js installation"
        exit 1      
fi
    fi
    check=$(git --version)
    if [ "$check" == "" ]; then
      "echo" "-e" "Installing Git"
      yum -y -q install git    
fi
    # Ubuntu
  else
    if [ "$distribution" == "ubuntu" ]; then
      "deb" 
      # Debian
    else
      if [ "$distribution" == "debian" ]; then
        "deb" 
      else
        # Unknown distribution
        error=$((0))
        check=$(git --version)
        if [ "$check" == "" ]; then
          error=$((1))
          "echo" "-e" "Error: git is missing"        
fi
        check=$(node -v)
        if [ "$check" == "" ]; then
          error=$((1))
          "echo" "-e" "Error: node is missing"        
fi
        if [ $(($error > 0)) == 1 ]; then
          "echo" "-e" "Please install above missing software"
          exit 1        
fi
      fi
    fi
  fi
  check=$(pm2 --version)
  if [ "$check" == "" ]; then
    "echo" "-e" "Installing PM2"
    npm install pm2 -g && pm2 install pm2-logrotate
    ln -s /usr/local/bin/pm2 /usr/bin/pm2
    pm2 startup  
fi
  mkdir -p $installPath
  cd $installPath
  git clone https://github.com/louislam/uptime-kuma.git .
  npm run setup
  pm2 start server/server.js --name uptime-kuma -- --port=$port
else
  defaultVolume="uptime-kuma"
  check=$(docker -v)
  if [ "$check" == "" ]; then
    "echo" "-e" "Error: docker is not found!"
    exit 1  
fi
  check=$(docker info)
  if [[ "$check" == *"Is the docker daemon running"* ]]; then
      "echo" "Error: docker is not running"
      "exit" "1"
    fi
  if [ "$3" != "" ]; then
    port="$3"
  else
    "read" "-p" "Expose Port [$defaultPort]: " "port"
    if [ "$port" == "" ]; then
      port="$defaultPort"    
fi
  fi
  if [ "$2" != "" ]; then
    volume="$2"
  else
    "read" "-p" "Volume Name [$defaultVolume]: " "volume"
    if [ "$volume" == "" ]; then
      volume="$defaultVolume"    
fi
  fi
  "echo" "-e" "Port: $port"
  "echo" "-e" "Volume: $volume"
  docker volume create $volume
  docker run -d --restart=always -p $port:3001 -v $volume:/app/data --name uptime-kuma louislam/uptime-kuma:1
fi
"echo" "-e" "http://localhost:$port"
